File: perl/eg/threaded.pl
#!/usr/bin/perl
use strict;
use warnings;
use threads;
use threads::shared;

use FCGI       qw[];
use IO::Handle qw[];

use constant THREAD_COUNT => 5;

my @count : shared = (0, (0) x THREAD_COUNT);

sub worker {
    my $k = shift;
    my %env;
    my $in  = IO::Handle->new;
    my $out = IO::Handle->new;
    my $err = IO::Handle->new;

    my $request = FCGI::Request($in, $out, $err, \%env);

    while ($request->Accept >= 0) {
        print $out
               "Content-type: text/html\r\n",
               "\r\n",
               "<title>FastCGI Hello! (multi-threaded perl, fcgiapp library)</title>",
               "<h1>FastCGI Hello! (multi-threaded perl, fcgiapp library)</h1>",
               "Request counts for ", THREAD_COUNT ," threads ",
               "running on host <i>$env{SERVER_NAME}</i>";

        {
            lock(@count);

            ++$count[$k];

            for(my $i = 1; $i <= THREAD_COUNT; $i++) {
                print $out $count[$i];
                print $out " ";
            }
        }
        $request->Flush;
        sleep(1);
    }
}

$_->join for map { threads->create(\&worker, $_) } 1..THREAD_COUNT;

File: examples/threaded.c
/*
 * threaded.c -- A simple multi-threaded FastCGI application.
 */

#ifndef lint
static const char rcsid[] = "$Id: threaded.c,v 1.2 1999/07/28 00:34:49 roberts Exp $";
#endif /* not lint */

#include "fcgi_config.h"

#include <pthread.h>

#ifdef HAVE_UNISTD_H
#include <unistd.h>
#endif

#ifdef _WIN32
#include <process.h>
#endif

#include "fcgiapp.h"


#define THREAD_COUNT 20

int count[THREAD_COUNT];

static void *doit(void *a)
{
    int i, k = (int)a;
    FCGX_Request request;
    FCGX_Stream *in, *out, *err;
    FCGX_ParamArray envp;
    char *server_name;

    FCGX_InitRequest(&request);

    while (FCGX_Accept_r(&in, &out, &err, &envp, &request) >= 0)
    {
        server_name = FCGX_GetParam("SERVER_NAME", envp);

        FCGX_FPrintF(out,
            "Content-type: text/html\r\n"
            "\r\n"
            "<title>FastCGI Hello! (multi-threaded C, fcgiapp library)</title>"
            "<h1>FastCGI Hello! (multi-threaded C, fcgiapp library)</h1>"
            "Thread %d, Request %d<p>"
            "Request counts for %d threads running on host <i>%s</i><p><code>",
            k, ++count[k], THREAD_COUNT, server_name ? server_name : "?");

        for (i = 0; i < THREAD_COUNT; i++)
            FCGX_FPrintF(out, "%5d " , count[i]);
    }

    return NULL;
}

int main(void)
{
    int i;
    pthread_t id[THREAD_COUNT];

    FCGX_Init();

    for (i = 0; i < THREAD_COUNT; i++)
        count[i] = 0;

    for (i = 1; i < THREAD_COUNT; i++)
        pthread_create(&id[i], NULL, doit, (void*)i);

    doit(0);

    exit(0);
}

